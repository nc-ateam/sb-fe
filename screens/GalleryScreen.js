import React from "react";
import { ImagePicker, Permissions } from "expo";
import { Image, View, Alert } from "react-native";
import * as firebase from "firebase";
import { NavigationBar, Icon, Title, Button, Text } from "@shoutem/ui";

class GalleryScreen extends React.Component {
  state = {
    image: null,
    photo_URL: ""
  };

  pickImage = async () => {
    const { status } = await Permissions.askAsync(Permissions.CAMERA_ROLL);

    if (status === "granted") {
      let result = await ImagePicker.launchImageLibraryAsync({
        allowsEditing: true,
        aspect: [4, 3],
        mediaTypes: "Images"
      });

      this.setState({ image: result.uri });
    }
  };

  uploadImage = async (uri, imageName) => {
    const { username, navigation, onRefresh } = this.props;
    const { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status === "granted") {
      const response = await fetch(this.state.image);
      let result = await Expo.Location.getCurrentPositionAsync();
      const blob = await response.blob();
      let filename = `~${result.coords.longitude},${
        result.coords.latitude
      }~${username}jpeg`;
      firebase
        .storage()
        .ref()
        .child(`${username}/` + filename)
        .put(blob)
        .then(response => {
          firebase
            .storage()
            .ref(username)
            .child(filename)
            .getDownloadURL()
            .then(url => {
              this.setState({ photo_URL: url }, () => {
                onRefresh();
                navigation.navigate("Cities");
              });
            });
        });
    }
  };

  componentDidUpdate(prevProps, prevState) {
    const { landmarkId } = this.props;
    if (prevState.photo_URL !== this.state.photo_URL) {
      console.log(this.state.photo_URL);
      fetch(
        `https://stamp-book-api.herokuapp.com/api/landmarks/${landmarkId}/checkLandmark`,
        {
          method: "POST",
          headers: {
            Accept: "application/json",
            "Content-Type": "application/json"
          },
          body: JSON.stringify({ body: this.state.photo_URL })
        }
      )
        .then(response => {
          console.log(response.status);
          // response.status - Error 400 for fail, 201 for pass.
          if (response.status === 400) {
            Alert.alert("You are not near the landmark!");
            return response.json();
          } else if (response.status === 201) {
            Alert.alert("Photo uploaded!");
            return response.json();
          } else if (response.status === 503) {
            Alert.alert(
              "Service is currently unavailable. Please try again later"
            );
            return response.json();
          }
        })
        .then(responseJson => {
          return responseJson.storedPhoto;
        })
        .catch(error => {
          console.error(error);
        });
    }
  }

  render() {
    let { image } = this.state;
    const { navigation } = this.props;
    return (
      <View
        style={{
          flex: 1,
          alignItems: "center",
          justifyContent: "center",
          backgroundColor: "#FFFFFF"
        }}
      >
        {image && (
          <Image source={{ uri: image }} style={{ width: 200, height: 200 }} />
        )}
        <Button
          style={{
            width: 140,
            marginTop: 20,
            marginBottom: 12,
            backgroundColor: "white",
            borderColor: "darkgrey"
          }}
          onPress={this.pickImage}
        >
          <Text style={{ color: "black" }}>Pick Image</Text>
        </Button>
        {this.state.image ? (
          <Button
            style={{
              width: 140,
              backgroundColor: "white",
              borderColor: "darkgrey"
            }}
            onPress={this.uploadImage}
          >
            <Text style={{ color: "black" }}>Upload Image</Text>
          </Button>
        ) : null}
        <NavigationBar
          style={{ container: { borderBottomColor: "#BDBDBD" } }}
          leftComponent={
            <Button onPress={() => navigation.openDrawer()}>
              <Icon name="sidebar" />
            </Button>
          }
          centerComponent={<Title>Gallery</Title>}
          rightComponent={
            <Button onPress={() => navigation.goBack()}>
              <Text style={{ color: "black", marginRight: 5 }}>Back</Text>
            </Button>
          }
        />
      </View>
    );
  }
}

export default GalleryScreen;
