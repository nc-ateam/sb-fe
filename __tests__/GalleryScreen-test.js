import React from "react";
import "react-native";
import GalleryScreen from "../screens/GalleryScreen";

import renderer from "react-test-renderer";

describe("<GalleryScreen />", () => {
  it("renders without crashing", () => {
    const rendered = renderer.create(<GalleryScreen />).toJSON();
    expect(rendered).toBeTruthy();
  });

  it("creates a screenshot", () => {
    const rendered = renderer.create(<GalleryScreen />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
