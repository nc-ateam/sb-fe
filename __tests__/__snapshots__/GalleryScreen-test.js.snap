// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GalleryScreen /> creates a screenshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onLayout={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderColor": "darkgrey",
        "borderRadius": 2,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 12,
        "marginTop": 20,
        "opacity": 1,
        "paddingLeft": 15,
        "paddingRight": 5,
        "width": 140,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      animationOptions={Object {}}
      ellipsizeMode="tail"
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "black",
          "fontFamily": "Rubik-Regular",
          "fontSize": 12,
          "fontStyle": "normal",
          "fontWeight": "600",
          "letterSpacing": 1,
          "marginBottom": 12,
          "marginRight": 10,
          "marginTop": 12,
        }
      }
    >
      Pick Image
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderBottomColor": "#BDBDBD",
        "borderBottomWidth": 1,
        "bottom": 0,
        "height": 70,
        "left": 0,
        "paddingTop": 20,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "backgroundColor": "transparent",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "center",
            "flex": 1,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#ffffff",
              "borderRadius": 0,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "opacity": 1,
              "paddingLeft": 9,
              "paddingRight": 9,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            animationOptions={Object {}}
            ellipsizeMode="tail"
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Object {
                  "backgroundColor": "transparent",
                  "color": "#222222",
                  "fontSize": 24,
                  "marginBottom": 9,
                  "marginRight": 0,
                  "marginTop": 9,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                Object {
                  "fontFamily": "rubicon-icon-font",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            î¤‰
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "flex": 1,
            "marginBottom": 0,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          animationOptions={Object {}}
          ellipsizeMode="tail"
          onLayout={[Function]}
          style={
            Object {
              "backgroundColor": "transparent",
              "color": "#222222",
              "fontFamily": "Rubik-Regular",
              "fontSize": 15,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 18,
            }
          }
        >
          Gallery
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "alignSelf": "center",
            "flex": 1,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#ffffff",
              "borderRadius": 0,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "opacity": 1,
              "paddingLeft": 9,
              "paddingRight": 9,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            animationOptions={Object {}}
            ellipsizeMode="tail"
            onLayout={[Function]}
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "black",
                "fontFamily": "Rubik-Regular",
                "fontSize": 15,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "letterSpacing": 0,
                "marginBottom": 12,
                "marginRight": 5,
                "marginTop": 12,
              }
            }
          >
            Back
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
